pkg/agent/memberlist/cluster_test.go

```diff
@@ -0,0 +1,643 @@
+					}
+				}
+			}
+			assert.Equal(t, expectEgressSeqSum, seqSum, "Egress seq")
+			for _, node := range nodes {
+				t.Logf("Node (%s) Egress: %#v", node, nodeEgress[node])
+			}
+			assert.Equal(t, egressNum, hitCount, "Egress total num should be 30")
+		})
+	}
+}
+
+func TestCluster_ShouldSelectEgress(t *testing.T) {
+	testCases := []struct {
+		name         string
+		nodeNum      int
+		egressIP     string
+		expectedNode string
+	}{
+		{
+			name:         "Select Node from 0 nodes",
+			nodeNum:      0,
+			egressIP:     "1.1.1.1",
+			expectedNode: "",
+		},
+		{
+			name:         "Select Node from 1 nodes",
+			nodeNum:      1,
+			egressIP:     "1.1.1.1",
+			expectedNode: "node-0",
+		},
+		{
+			name:         "Select Node from 3 nodes",
+			nodeNum:      3,
+			egressIP:     "1.1.1.1",
+			expectedNode: "node-1",
+		},
+		{
+			name:         "Select Node from 10 nodes",
+			nodeNum:      10,
+			egressIP:     "1.1.1.1",
+			expectedNode: "node-1",
+		},
+		{
+			name:         "Select Node from 100 nodes",
+			nodeNum:      100,
+			egressIP:     "1.1.1.1",
+			expectedNode: "node-79",
+		},
+	}
+	for _, tCase := range testCases {
+		t.Run(tCase.name, func(t *testing.T) {
+			genLocalNodeCluster := func(nodeNme string) *Cluster {
+				cluster := &Cluster{
+					mList:             &memberlist.Memberlist{},
+					nodeName:          nodeNme,
+					consistentHashMap: make(map[string]*consistenthash.Map),
+				}
+				return cluster
+			}
+
+			genNodes := func(n int) []string {
+				nodes := make([]string, n)
+				for i := 0; i < n; i++ {
+					nodes[i] = fmt.Sprintf("node-%d", i)
+				}
+				return nodes
+			}
+
+			fakeEIPName := "fakeExternalIPPool"
+			fakeEgress := &crdv1a2.Egress{
+				ObjectMeta: metav1.ObjectMeta{Name: "fakeEgress"},
+				Spec:       crdv1a2.EgressSpec{ExternalIPPool: fakeEIPName, EgressIP: tCase.egressIP},
+			}
+
+			if tCase.nodeNum == 0 {
+				fakeCluster := genLocalNodeCluster("fakeLocalNodeName")
+				fakeCluster.consistentHashMap[fakeEIPName] = consistenthash.New(defaultVirtualNodeReplicas, nil)
+				selected, err := fakeCluster.ShouldSelectEgress(fakeEgress)
+				assert.NoError(t, err)
+				assert.Equal(t, false, selected, "Select Node for Egress not match")
+			} else {
+				nodes := genNodes(tCase.nodeNum)
+				var actualNodes []string
+				for _, node := range nodes {
+					fakeCluster := genLocalNodeCluster(node)
+					fakeCluster.consistentHashMap[fakeEIPName] = consistenthash.New(defaultVirtualNodeReplicas, nil)
+					fakeCluster.consistentHashMap[fakeEIPName].Add(nodes...)
+					selected, err := fakeCluster.ShouldSelectEgress(fakeEgress)
+					assert.NoError(t, err)
+					if selected {
+						actualNodes = append(actualNodes, node)
+						t.Logf("Node %s selected from %d Nodes", node, tCase.nodeNum)
+					}
+				}
+				assert.Equal(t, 1, len(actualNodes), "Selected Node num for Egress not match")
+				assert.Equal(t, []string{tCase.expectedNode}, actualNodes, "Select Node for Egress not match")
```

Comment:
The test itself is O(n^2), I understand it needs to construct a fakeCluster for each Node to check whether a Node can select the Egress in current implementation. Why don't just change `ShouldSelectNode` return the desired Node to make the test much easier. The client just needs to check whether the desired Node is itself.

Reply:
done